### PR TITLE
add state to socketLink to prevent errors

### PR DESCRIPTION
**Pull request by brycekahle**
### What does this PR do?

Adds `attached` state to `socketLink`

### Motivation

If you called `Pause()` then `Close()`, it would result in an error

### CODE REVIEW

File: socket.go [hunk 1]
```diff
@@ -1,41 +1,76 @@
...
 
 	"golang.org/x/sys/unix"
 )
 
 // attachSocket - Attaches the probe to the provided socket
-func (p *Probe) attachSocket() error {
-	if err := sockAttach(p.SocketFD, p.program.FD()); err != nil {
-		return err
+func (p *Probe) attachSocket() (err error) {
+	p.progLink, err = newSocketLink(p.SocketFD, p.program.FD())
```
Comment (mbertrone) [inline socket.go]: issue (blocking): on attach failure this stores a typed-nil `*socketLink` into `p.progLink`, which then panics during cleanup. `newSocketLink` returns `(nil, err)` when `sl.attach()` fails, but that nil is a `*socketLink`, so assigning it to the `io.Closer` field (`probe.go:55`) yields a non-nil interface wrapping a nil pointer. `Probe.attach()` runs `p.stop(false)` on the error path, and `Probe.detach()` (`probe.go:702`) does `if p.progLink != nil { ... p.progLink.Close() }`. The guard passes because the interface is non-nil, `Close()` → `detach()` → `s.mtx.Lock()` dereferences the nil receiver, and `internal.Retry` has no recover, so the process crashes.

This is a regression from the old code, which returned before assigning `p.progLink` when `sockAttach` failed. Fix by not publishing the link until it's live:
```go
func (p *Probe) attachSocket() error {
	sl, err := newSocketLink(p.SocketFD, p.program.FD())
	if err != nil {
		return err
	}
	p.progLink = sl
	return nil
}
```

File: socket.go [hunk 2]
```diff
@@ -1,41 +1,76 @@
...
+	defer s.mtx.Unlock()
+	if !s.attached {
+		return nil
+	}
+
+	if err := syscall.SetsockoptInt(s.sockFD, syscall.SOL_SOCKET, unix.SO_DETACH_BPF, s.progFD); err != nil {
+		if errors.Is(err, unix.ENOENT) {
+			s.attached = false
+		}
+		return err
```
Comment (mbertrone) [inline socket.go:50]: suggestion (non-blocking): the `ENOENT` branch reconciles state to `attached=false` but still returns the error, so an already-detached socket surfaces a spurious error through `Close()`/`Pause()` into the joined result at `probe.go:703`. Not a regression, and only reachable when kernel and tracked state diverge, but returning the error here works against the idempotent-detach intent of this PR. If that divergence is the expected "already detached" case, return nil after clearing the flag:
```go
if errors.Is(err, unix.ENOENT) {
	s.attached = false
	return nil
}
return err
```

File: socket.go (hunk 1 above)
Comment (brycekahle) [reply socket.go]: Fixed.

File: socket.go (hunk 2 above)
Comment (brycekahle) [reply socket.go:50]: I think surfacing the error when kernel and tracked state diverge is important

**Review by mbertrone — APPROVED** [review]
